test/semantics/omp-clause-validity01.f90

```diff
@@ -206,6 +206,26 @@
   do i = 1, N
      a = 3.14
   enddo
+  !$omp end parallel do
+
+! Loop association check
+
+  a = 0.0
+  !ERROR: The END PARALLEL DO must follow the do-loop associated with the loop construct
+  !$omp end parallel do
+  !$omp parallel do private(c)
+  do i = 1, N
+     do j = 1, N
+        !ERROR: do-loop is expected after the PARALLEL DO directive
+        !$omp parallel do shared(b)
+        a = 3.14
+     enddo
+     !ERROR: The END PARALLEL DO must follow the do-loop associated with the loop construct
+     !$omp end parallel do
+  enddo
+  a = 1.414
+  !ERROR: The END PARALLEL DO must follow the do-loop associated with the loop construct
+  !$omp end parallel do
```

Comment:
PARALLEL do should work on a labeled DO loop right? You should have a test for that.

Also, what happens if the directive is on a DO CONCURRENT?

Reply:
Good question. I thought DO CONCURRENT was regular DoConstruct. Let me try.

Reply:
Yep, DO CONCURRENT is DoConstruct in the parse tree. It works with OpenMP.

Reply:
Is DO CONCURRENT documented in the OpenMP spec?

Reply:
You might want to be selective about allowing a DO CONCURRENT loop as a OpenMP loop construct.  The Fortran standard knows nothing about OpenMP.  OpenMP knows/says nothing about DO CONCURRENT.  That is because no one has figured out what to do with this combination.  One option is to disallow OpenMP directives on a DO CONCURRENT loop.  Another is to selectively allow some simple cases.  Another is to take the time to allow as many cases as possible.

Reply:
pgi fortran currently implements the 2nd of these three choices.

Reply:
OpenMP 4.5 spec doesn't mention anything about DO CONCURRENT, yet 5.0 spec says:

> This OpenMP API specification refers to ISO/IEC 1539-1:2010 as Fortran 2008. While future
> versions of the OpenMP specification are expected to address the following features, currently
> their use may result in unspecified behavior.
>   – Submodules
>   – Coarrays
>   – DO CONCURRENT
>   ...

GNU doesn't allow OpenMP + DO CONCURRENT but our PGI compiler does.

Reply:
> pgi fortran currently implements the 2nd of these three choices.

Good to know.

Reply:
To be more specific,  flang / pgi fortran has code that does DO CONCURRENT LOCAL_INIT assignments and mask expressions, regardless of OpenMP.  In the absence of compiler parallelization of a DO CONCURRENT, OpenMP processing will otherwise ignore locality-spec information.  If there are no locality-spec clauses,  a DO CONCURRENT loop will look like a loop nest to OpenMP, and OpenMP clauses like COLLAPSE should work.  There is no attempt to treat locality-spec clause as PRIVATE, FIRSTPRIVATE, or SHARED clauses.